src/pamiq_core/trainer/base.py

```diff
@@ -0,0 +1,122 @@
+from abc import ABC, abstractmethod
+from typing import Any
+
+from pamiq_core.data import DataUser, DataUsersDict
+from pamiq_core.model import (
+    InferenceModelsDict,
+    TrainingModel,
+    TrainingModelsDict,
+)
+
+
+class Trainer(ABC):
+    """Abstract base trainer class.
+
+    The `run` method is called repeatedly in the training thread.
+
+    Override the following methods:
+        - `on_training_models_attached`: Callback method for when `model_wrappers_dict` is attached to the trainer.
+        - `on_data_users_dict_attached`: Callback method when `data_users_dict` is attached to the trainer.
+        - `is_trainable`: Return whether the training can be executed.
+        - `setup`: To setup before training starts.
+        - `train`: The training process.
+        - `teardown`: To teardown after training.
+
+    Models and data buffers become available after the thread has started.
+    """
+
+    _training_models_dict: TrainingModelsDict
+    _data_users_dict: DataUsersDict
+    _retrieved_model_names: set[str]
+
+    def __init__(self) -> None:
+        """Initialize."""
+        super().__init__()
+        self._retrieved_model_names = set()
```

Comment:
コンストラクタで設定する変数はクラスフィールドにアノテーションをつけなくて良い。あくまでコンストラクタでは設定されない、メンバー変数を列挙しておく。
```suggestion

    def __init__(self) -> None:
        """Initialize."""
        super().__init__()
        self._retrieved_model_names: set[str] = set()
```